url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/util/SourceGenerator.java

```diff
@@ -0,0 +1,216 @@
+		<li>Turn ParserTables (built from a Syntax) into compileable Java-code (for faster loading).</li>
+		<li>Turn a Syntax object into compileable Java code (for faster loading).</li>
+	</ul>
+	<pre>
+	SYNTAX: java fri.patterns.interpreter.parsergenerator.util.SourceGenerator [semantic|LALR|SLR|LR] file.syntax [file.syntax ...]
+			LALR|SLR|LR: Generates ParserTable implementation(s) of passed grammar file(s).
+			       else: Generates Syntax implementation(s) of passed grammar file(s).
+	CAUTION: Files MUST have relative pathes!
+	</pre>
+
+	@author (c) 2002, Fritz Ritzberger
+*/
+
+public abstract class SourceGenerator
+{
+	/**
+		Generates a semantic skeleton implementation for a given syntax.
+		@param input syntax the semantic is meant for
+		@param className basename of class to generate, semantic will be named className+"Semantic.java"
+		@param pkgName package-name of class to generate
+	*/
+	public static void generateSemanticSkeleton(Syntax syntax, String className, String pkgName)
+		throws Exception
+	{
+		String fileName = className+"Semantic.java";
+
+		String dirName = pkgName != null && pkgName.length() > 0
+				? pkgName.replace('.', File.separatorChar)
+				: System.getProperty("user.dir");
+
+		File out = new File(dirName, fileName);
+				
+		if (out.exists())	{
+			throw new IllegalStateException("Will not overwrite "+out.getAbsolutePath()+". Please check the file for implementation and remove it!");
+		}
+		else	{
+			new File(dirName).mkdirs();	// ensure directory exists
+			FileWriter fw = new FileWriter(out);
+			new SemanticSkeletonGenerator(syntax, className, pkgName, fw);
+			System.err.println("Wrote semantic skeleton to file: "+out);
+		}
+	}
+
+
+	/**
+		Generates Java code from parser tables. makes the output filename and calls <i>parserTables.toSourceFile()</i> then.
+		@param parserTables ParserTables for which the source should stand
+		@param className basename of class to generate
+		@param pkgName name of package of class to generate, can be null
+	*/
+	public static void generateParserTable(AbstractParserTables parserTables, String className, String pkgName)
+		throws Exception
+	{
+		String fullName = (pkgName != null ? pkgName+"." : "")+className;
+		parserTables.toSourceFile(fullName);
+	}
+
+
+	/**
+		Generates a Java implementation from the passed Syntax object.
+		@param syntax Syntax to convert to Java code.
+		@param className basename of class to generate
+		@param pkgName name of package of class to generate
+	*/
+	public static void generateSyntaxImpl(
+		Syntax syntax,
+		String className,
+		String pkgName,
+		List initialNonterminals)
+		throws IOException
+	{
+		String origClsName = className;
+		className = className+"Syntax";
+		String fileName = (pkgName != null ? pkgName+"."+className : className);
+		fileName = fileName.replace('.', File.separatorChar)+".java";
+		Writer f = new BufferedWriter(new FileWriter(fileName));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *f*. The resource is referenced at lines 89, 92, 94, 95, 96, 97, 98, 100, 101, 102, 105, 106, 108, 109, 110, 117, 122, 125, 127, 129, 131, 132. The resource is closed at line 134. There are other execution paths that do not contain closure statements, for example, when Writer.write (in method fwrite) throws an exception. Ensure *f* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).